File: src/shipit/procfile.py
# File forked from https://github.com/nickstenning/honcho/blob/main/honcho/environ.py
# MIT License
import re

PROCFILE_LINE = re.compile(r"^([A-Za-z0-9_-]+):\s*(.+)$")


class Procfile(object):
    """A data structure representing a Procfile"""

    def __init__(self):
        self.processes = {}

    @classmethod
    def loads(cls, contents):
        p = cls()
        for line in contents.splitlines():
            m = PROCFILE_LINE.match(line)
            if m:
                p.add_process(m.group(1), m.group(2))
        return p

    def add_process(self, name, command):
        assert name not in self.processes, (
            "process names must be unique within a Procfile"
        )
        self.processes[name] = command

    def get_start_command(self):
        if "web" in self.processes:
            return self.processes["web"]
        elif "default" in self.processes:
            return self.processes["default"]
        elif "start" in self.processes:
            return self.processes["start"]
        elif len(self.processes) == 1:
            return list(self.processes.values())[0]
        return None

File: crates/shipit-providers/src/procfile.rs
//! Procfile parsing (port of `procfile.py`, itself forked from honcho).

use anyhow::{bail, Result};
use indexmap::IndexMap;

pub struct Procfile {
    pub processes: IndexMap<String, String>,
}

impl Procfile {
    pub fn loads(contents: &str) -> Result<Self> {
        let line_re = regex::Regex::new(r"^([A-Za-z0-9_-]+):\s*(.+)$").unwrap();
        let mut processes = IndexMap::new();
        for line in contents.lines() {
            if let Some(captures) = line_re.captures(line) {
                let name = captures[1].to_owned();
                let command = captures[2].to_owned();
                if processes.contains_key(&name) {
                    bail!("process names must be unique within a Procfile");
                }
                processes.insert(name, command);
            }
        }
        Ok(Self { processes })
    }

    pub fn get_start_command(&self) -> Option<&str> {
        for key in ["web", "default", "start"] {
            if let Some(command) = self.processes.get(key) {
                return Some(command);
            }
        }
        if self.processes.len() == 1 {
            return self.processes.values().next().map(String::as_str);
        }
        None
    }
}
